### PR TITLE
Booking and payment service

### PR DESCRIPTION
**Pull request by truongtamthanh2004**
# ✅ Checklist Tự review
## 🧹 Code sạch & format chuẩn
- [x] Code sạch, dễ đọc, theo convention của team (naming, cấu trúc)
- [ ] Đã chạy `go fmt`, `goimports`, `golangci-lint` (nếu có)
- [x] Không còn đoạn code tạm, `fmt.Println`, `log.Println`, comment debug
- [x] Không còn biến, hàm, import không sử dụng
- [x] Hàm không quá dài, có thể tách nhỏ nếu vượt quá 50 dòng
- [x] Code không lặp lại, đã tái sử dụng hợp lý

## 🧠 Logic & Xử lý lỗi
- [x] Logic đúng yêu cầu, không dư thừa, rõ ràng
- [x] Xử lý đầy đủ các trường hợp `nil`, `error`, `invalid input`
- [x] Mỗi thao tác có thể lỗi (I/O, DB, parse...) đều được check `err`
- [x] Không để `err` bị silent — luôn log hoặc trả về (tùy ngữ cảnh)
- [x] Sử dụng `context` đúng cách (nếu dùng), tránh leak
- [x] Nếu là `goroutine`, có quản lý lifecycle và tránh leak

## 🌐 HTTP Handler (Gin)
- [x] Sử dụng đúng HTTP method (GET, POST, PUT, DELETE…)
- [x] Validate request (param, query, body) đầy đủ, rõ ràng
- [x] Sử dụng binding (`ShouldBindJSON`, `ShouldBindUri`, v.v.) đúng cách
- [x] Xử lý trả về `status code` phù hợp (`200`, `201`, `400`, `404`, `500`…)
- [x] Response rõ ràng, có `message`, `data`, `error` (nếu cần)
- [x] Không leak thông tin nhạy cảm trong response lỗi

## 🗃️ Struct & Layer
- [x] Struct có annotation JSON đầy đủ (`json:"field_name"`)
- [x] Sử dụng DTO/input/output rõ ràng, không lẫn lộn model DB và response
- [x] Đã tách rõ các layer: handler → service → repository (nếu theo clean code)
- [x] Middleware viết rõ ràng, dễ đọc, không làm chậm hệ thống

## 🧪 Unit Test & Mock
- [x] Có test cho handler/service chính (ít nhất happy case và edge case)
- [x] Nếu dùng DB/service ngoài, có mock hoặc test isolation
- [x] Có test validate đầu vào (request, param, query)
- [x] Không dùng test hardcode giá trị không ổn định (vd: time.Now() mà không mock)
- [x] Tên test rõ ràng: `TestFunction_Scenario_Expected`

## 🔐 Security & Config
- [x] Không hardcode secret, token, config trong code
- [x] Không để lộ thông tin nhạy cảm qua log hoặc panic
- [x] Validate và sanitize đầu vào người dùng
- [x] Nếu có auth, kiểm tra middleware hoạt động đúng

## 🤖 Sử dụng GitHub Copilot
- [x] Đã kiểm tra kỹ mọi đoạn code được gợi ý từ GitHub Copilot
- [x] Không để lại phần comment hoặc TODO do Copilot sinh ra
- [x] Không commit đoạn code mà chưa hiểu rõ logic hoặc mục đích
- [x] Đảm bảo mọi function/code block từ Copilot phù hợp với context dự án
- [x] Tối ưu lại code được gợi ý nếu cần — tránh lạm dụng hoặc copy nguyên mẫu

## Related Tickets
- ticket redmine link

## WHAT (optional)
- Change number items `completed/total` in admin page.

## HOW
- I edit js file, inject not_vary_normal items in calculate function.

## WHY (optional)
- Because in previous version - number just depends on `normal` items. But in new version, we have `state` and `confirm_state` depends on both `normal` + `not_normal` items.

### CODE REVIEW

**Review by ledat139 — APPROVED** [review]

**Comment by ledat139** [comment]
READY

**Comment by truongtamthanh2004** [comment]
ready

**Comment by truongtamthanh2004** [comment]
ready

File: services/booking-service/internal/handler/booking_handler.go [hunk 1]
```diff
@@ -1 +1,137 @@
...
+	roleVal, _ := c.Get("role")
+	role := roleVal.(string)
+	if role != "admin" && role != "moderator" {
+		c.JSON(http.StatusForbidden, gin.H{"message": "forbidden"})
+		return
+	}
+
+	bookings, err := h.usecase.GetAllBooking()
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{"message": err.Error()})
```
Comment (chienpv-3590) [inline services/booking-service/internal/handler/booking_handler.go:132]: Khi gọi các hàm usecase, nếu lỗi đến từ lỗi nghiệp vụ (ví dụ: không thể đặt phòng vì hết chỗ), nên trả về mã lỗi phù hợp hơn là InternalServerError.

File: services/booking-service/internal/usecase/booking_usecase.go [hunk 2]
```diff
@@ -1 +1,116 @@
...
+	err = u.repo.Update(booking)
+	if err != nil {
+		return nil, err
+	}
+
+	if u.producer != nil {
+		event := map[string]interface{}{
+			"user_id": booking.UserID,
+			"type":    "BOOKING_STATUS_UPDATED",
+			"content": fmt.Sprintf("Your booking %d status changed to %s", booking.ID, booking.Status),
```
Comment (chienpv-3590) [inline services/booking-service/internal/usecase/booking_usecase.go:93]: Khi cập nhật status, chưa kiểm tra giá trị status hợp lệ (ví dụ: chỉ cho phép chuyển sang CONFIRMED, CANCELLED,...).

File: services/booking-service/internal/usecase/booking_usecase.go [hunk 3]
```diff
@@ -1 +1,116 @@
...
+
+	totalPrice := duration * space.Price
+
+	booking := &model.Booking{
+		UserID:     userID,
+		SpaceID:    spaceID,
+		StartTime:  start,
+		EndTime:    end,
+		TotalPrice: totalPrice,
+		Status:     "PENDING",
```
Comment (chienpv-3590) [inline services/booking-service/internal/usecase/booking_usecase.go:55]: Status booking đang hardcode ("PENDING"), nên define thành constant để tránh lỗi chính tả và thuận tiện refactor.

File: services/booking-service/internal/handler/booking_handler.go (hunk 1 above)
Comment (truongtamthanh2004) [reply services/booking-service/internal/handler/booking_handler.go:132]: này là hàm getAll nên em nghĩ không có thêm logic gì cả anh ạ

**Comment by truongtamthanh2004** [comment]
ready
